pkg/agent/controller/networkpolicy/packetin.go

```diff
@@ -327,3 +335,101 @@ func (c *Controller) rejectRequest(pktIn *ofctrl.PacketIn) error {
 			true)
 	}
 }
+
+func (c *Controller) storeDenyConnection(pktIn *ofctrl.PacketIn) error {
+
+	packet, err := binding.ParsePacketIn(pktIn)
+	if err != nil {
+		return fmt.Errorf("error in parsing packetin: %v", err)
+	}
+
+	denyConn, err := c.generateDenyConn(pktIn, packet)
+	if err != nil {
+		return fmt.Errorf("error when getting deny connection information: %v", err)
+	}
+	c.denyConnectionStore.AddOrUpdateConnection(denyConn)
+	return nil
+}
+
+func (c *Controller) generateDenyConn(pktIn *ofctrl.PacketIn, packet *binding.Packet) (*flowexporter.DenyConnection, error) {
+	denyConn := flowexporter.DenyConnection{}
+
+	// get 5-tuple information
+	flowKey := flowexporter.Tuple{
+		SourcePort:      packet.SourcePort,
+		DestinationPort: packet.DestinationPort,
+		Protocol:        packet.IPProto,
+	}
+	if packet.IsIPv6 {
+		flowKey.SourceAddress = packet.SourceIP.To16()
+		flowKey.DestinationAddress = packet.DestinationIP.To16()
+		denyConn.IsIPv6 = true
+	} else {
+		flowKey.SourceAddress = packet.SourceIP.To4()
+		flowKey.DestinationAddress = packet.DestinationIP.To4()
+	}
+	denyConn.FlowKey = flowKey
+
+	// map local source IP and destination IP
+	sIface, srcFound := c.ifaceStore.GetInterfaceByIP(flowKey.SourceAddress.String())
+	dIface, dstFound := c.ifaceStore.GetInterfaceByIP(flowKey.DestinationAddress.String())
+	node, _ := env.GetNodeName()
+	if !srcFound && !dstFound {
+		klog.Warningf("Cannot map any of the IP %s or %s to a local Pod", flowKey.SourceAddress.String(), flowKey.DestinationAddress.String())
+	}
+	if srcFound && sIface.Type == interfacestore.ContainerInterface {
+		denyConn.SourcePodName = sIface.ContainerInterfaceConfig.PodName
+		denyConn.SourcePodNamespace = sIface.ContainerInterfaceConfig.PodNamespace
+		denyConn.SourceNodeName = node
+	}
+	if dstFound && dIface.Type == interfacestore.ContainerInterface {
+		denyConn.DestinationPodName = dIface.ContainerInterfaceConfig.PodName
+		denyConn.DestinationPodNamespace = dIface.ContainerInterfaceConfig.PodNamespace
+		denyConn.DestinationNodeName = node
+	}
+
```

Comment:
We cannot add service info because of the unavailability of the connection mark. Is that right? 
We can still use the destination IP and see if we can find a match in the proxy's service map right? If there is no match, then we do not add it because it is not a Pod-To-Service flow. Could we do that?

Reply:
Yes, we cannot get the connection mark. Makes sense to me. Updated.